### DISCUSSION TITLE
Merge Mylyn repositories

### DISCUSSION THREAD

**Discussion by ruspl-afed** [eclipse-mylyn/.github#13] [Polls]
Currently we have a number of git repositories for Mylyn components
* [commons](https://github.com/eclipse-mylyn/org.eclipse.mylyn.commons)
* [tasks](https://github.com/eclipse-mylyn/org.eclipse.mylyn.tasks)
* [contexts](https://github.com/eclipse-mylyn/org.eclipse.mylyn.context)
* [versions](https://github.com/eclipse-mylyn/org.eclipse.mylyn.versions)
* [builds](https://github.com/eclipse-mylyn/org.eclipse.mylyn.builds)
* [reviews](https://github.com/eclipse-mylyn/org.eclipse.mylyn.reviews)

Do we really need to have them all separated?
* In fact, they are not independent and have circular dependencies
* More repositories require much more releng effort
* Code changes require extra work to be aligned accross several repositories

How about merging them all to one? Of cause, after resolving  https://github.com/eclipse-mylyn/.github/issues/1

**Comment by BeckerFrank** [comment]
Maybe org.eclipse.mylyn.commons is an exception. 
org.eclipse.mylyn.docs use direct or indirect
- org.eclipse.mylyn.commons.context
- org.eclipse.mylyn.commons.core
- org.eclipse.mylyn.commons.screenshots
- org.eclipse.mylyn.commons.ui
- org.eclipse.mylyn.commons.workbench
- org.eclipse.mylyn.monitor.core
- org.eclipse.mylyn.monitor.ui

so it can make sense to have these in an separate repository.

**Reply by wimjongman** [reply]
What would be the benefit for them, Frank?

**Reply by ruspl-afed** [reply]
Frank, this use case is handled by having a dedicated feature published to Mylyn p2 repository, it does not require to have dedicated GitHub repository.

**Reply by BeckerFrank** [reply]
> Frank, this use case is handled by having a dedicated feature published to Mylyn p2 repository, it does not require to have dedicated GitHub repository.

sure, but having a separate repository can make it easier  for maintaining from both sides.
We can start with a  dedicated feature and if it is needed we can create a new repository later.

**Reply by wimjongman** [reply]
Maybe @akurtakov can chime in?

**Reply by akurtakov** [reply]
From releng POV having these multiple repositories is a total nightmare. Esp. considering the circular dependencies and how not test suite can reliably test these. Thus Eclipse Platform is merging 1-2 repos together for each release - this had no negative effect so far even the verification builds are not slower.
Regarding Mylyn Docs - nothing from mylyn commons is used anymore in the last releases, the code is there but it's not compiled and not part of the release.

**Reply by akurtakov** [reply]
IIRC this was the only way to continue releasing when Mylyn was removed from Simrel and no one complained since then about the lack of task-focused editing for wiki files so right now there are no plans to bring this functionality back even when Mylyn is back in simrel.

**Reply by wimjongman** [reply]
> Regarding Mylyn Docs - nothing from mylyn commons is used anymore

Cheers! So no block to merge commons in as well.

> lack of task-focused editing for wiki files so right now there are no plans to bring this functionality back even when Mylyn is back in simrel

Sounds totally valid. We will see if there is community interest in that once Mylyn is back on track.

**Comment by wimjongman** [comment]
What will the repo structure be? 

```
commons
tasks
etc..??
```
or one flat repo?

**Reply by ruspl-afed** [reply]
Pesonally I would prefer one flat, but do not see problems having subfolders

**Reply by wimjongman** [reply]
The problem with subfolders is that you cannot go tycho pomless (AFAIK)
The advantage of subfolders is that we have some organization

**Reply by ruspl-afed** [reply]
> is that you cannot go tycho pomless

I think we can, but with a bit more extra manifests. However, 

I thought you will offer a "simplicity" of merging as a reason to have subfolders.

**Reply by wimjongman** [reply]
Yes, there is that too. And the irresistible human urge to categorize. ;)

**Reply by BeckerFrank** [reply]
> Yes, there is that too. And the irresistible human urge to categorize. ;)

Somtimes categorize can help but we can do this with the Oomph *.setup and Working Sets. At some point we maybe want to know what Eclipse Project was in a specific Repository.

**Comment by wimjongman** [comment]
Guys, I have merged the repositories into the org.eclipse.mylyn repo retaining all history. I have created a new branch **_main_**, which I suggest we use going forward.

Please take a look at https://github.com/eclipse-mylyn/org.eclipse.mylyn/tree/main

I have created a bash script (attached) that I assembled from [this SO answer](https://stackoverflow.com/questions/13040958/merge-two-git-repositories-without-breaking-file-history/20974621#20974621). This means that I can replay it if something is missing/wrong.

Let me know if you miss something

[mylyn.sh.txt](https://github.com/eclipse-mylyn/.github/files/10295362/mylyn.sh.txt)

**Reply by wimjongman** [reply]
The reasons that I used subdirectories in the end is that this avoids filename clashes. For example pom.xml but this avoids also directoy name clashes should they occur.

**Reply by ruspl-afed** [reply]
Let's assume it is good. What would be the next step?

**Reply by BeckerFrank** [reply]
> The reasons that I used subdirectories in the end is that this avoids filename clashes. For example pom.xml but this avoids also directoy name clashes should they occur.

Do we really want to include org.eclipse.mylyn.docs in the new main branch?
In the past, Mylyn Docs was maintained completely separately!

**Reply by ruspl-afed** [reply]
Great catch, Frank! Mylyn Docs should not be included.

**Reply by wimjongman** [reply]
Why not? It is an active repository. @akurtakov WDYT?

**Reply by wimjongman** [reply]
> Let's assume it is good. What would be the next step?

Use it in the build and see if it works.

**Reply by ruspl-afed** [reply]
For me the main reason to have separate git repo is different lifecycle (/schedule/maturity) for the code. 
Mylyn Docs seems to have its own life already while all the rest are quite coupled. Moving Mylyn Docs back to overs looks a bit artificial for me. But let's wait for other opinions.

**Reply by ruspl-afed** [reply]
> > Let's assume it is good. What would be the next step?
> 
> Use it in the build and see if it works.

Just [fixed](https://github.com/eclipse-mylyn/.github/issues/1) all the CI for `master` branches . So, the pre-requisite of merge is already delivered.

**Reply by leods** [reply]
@ruspl-afed is correct, Mylyn Docs should not be merged into this repository.

**Reply by BeckerFrank** [reply]
> For me the main reason to have separate git repo is different lifecycle (/schedule/maturity) for the code.
> Mylyn Docs seems to have its own life already while all the rest are quite coupled. Moving Mylyn Docs back to overs looks a bit artificial for me. But let's wait for other opinions.

For me we have actual two independent parts in eclipse-mylyn so we should have two separate Repositories but we must make it clear in the main README.md so that Mylyn Docs related Discussions, Issues, Pull Requests are filled in the correct Repository.

All other Repositories should be Archived.

**Reply by ruspl-afed** [reply]
> Use it in the build and see if it works.

Do you want me to fix the build right in the `main` branch of `https://github.com/eclipse-mylyn/org.eclipse.mylyn/`, Wim? 
I'm asking again, just to avoid misunderstandings.

**Reply by BeckerFrank** [reply]
The Mylyn.setup and the CONTRIBUTING.md are now changed the the use main instead of master.

I think the next steps are.

- [x] Request to change the Default Branch from master to main for repository org.eclipse.mylyn
- [ ] If we want to have a review for every change we also should request that the main Branch  of org.eclipse.mylyn is protected
- [x] Request archive for the following repositories.
    - org.eclipse.mylyn.commons
    - org.eclipse.mylyn.builds
    - org.eclipse.mylyn.context
    - org.eclipse.mylyn.tasks
    - org.eclipse.mylyn.reviews
    - org.eclipse.mylyn.versions
    - org.eclipse.mylyn.all
    - org.eclipse.mylyn.incubator
    - org.eclipse.mylyn.reviews.ui
    - org.eclipse.mylyn.context.mft

Thougths?

**Reply by ruspl-afed** [reply]
+1 
Let's also archive 
- org.eclipse.mylyn.all

**Reply by BeckerFrank** [reply]
> +1 Let's also archive
> 
> * org.eclipse.mylyn.all

I add some more!

**Reply by BeckerFrank** [reply]
I create https://gitlab.eclipse.org/eclipsefdn/helpdesk/-/issues/2445 for the change from master to main
and
https://gitlab.eclipse.org/eclipsefdn/helpdesk/-/issues/2446 for Retirement of the repositories.
